service/matching/matching_engine_test.go

```diff
@@ -2761,6 +2762,321 @@ func (s *matchingEngineSuite) TestUpdatePhysicalTaskQueueGauge_BuildID() {
 
 }
 
+// generateWorkflowExecution is a helper to make a sample workflowExecution and WorkflowType for the required tests
+func (s *matchingEngineSuite) generateWorkflowExecution() (*commonpb.WorkflowType, *commonpb.WorkflowExecution) {
+	runID := uuid.NewRandom().String()
+	workflowID := "workflow1"
+	workflowType := &commonpb.WorkflowType{
+		Name: "workflow",
+	}
+	workflowExecution := &commonpb.WorkflowExecution{RunId: runID, WorkflowId: workflowID}
+	return workflowType, workflowExecution
+}
+
+func (s *matchingEngineSuite) mockHistoryWhilePolling(workflowType *commonpb.WorkflowType) {
+	s.mockHistoryClient.EXPECT().RecordWorkflowTaskStarted(gomock.Any(), gomock.Any(), gomock.Any()).DoAndReturn(
+		func(ctx context.Context, taskRequest *historyservice.RecordWorkflowTaskStartedRequest, arg2 ...interface{}) (*historyservice.RecordWorkflowTaskStartedResponse, error) {
+			return &historyservice.RecordWorkflowTaskStartedResponse{
+				PreviousStartedEventId: 1,
+				StartedEventId:         1,
+				ScheduledEventId:       1,
+				WorkflowType:           workflowType,
+				Attempt:                1,
+				History:                &historypb.History{Events: []*historypb.HistoryEvent{}},
+				NextPageToken:          nil,
+			}, nil
+		}).AnyTimes()
+}
+
+func (s *matchingEngineSuite) createTQAndPTQForBacklogTests() (*taskqueuepb.TaskQueue, *PhysicalTaskQueueKey) {
+	tq := "approximateBacklogCounter"
+	ptq := newUnversionedRootQueueKey(namespaceId, tq, enumspb.TASK_QUEUE_TYPE_WORKFLOW)
+	s.taskManager.getQueueManager(ptq).rangeID = 1
+	s.matchingEngine.config.RangeSize = 10
+
+	taskQueue := &taskqueuepb.TaskQueue{
+		Name: tq,
+		Kind: enumspb.TASK_QUEUE_KIND_NORMAL,
+	}
+
+	return taskQueue, ptq
+}
+
+func (s *matchingEngineSuite) addWorkflowTask(workflowExecution *commonpb.WorkflowExecution, taskQueue *taskqueuepb.TaskQueue) {
+	addRequest := matchingservice.AddWorkflowTaskRequest{
+		NamespaceId:            namespaceId,
+		Execution:              workflowExecution,
+		ScheduledEventId:       1,
+		TaskQueue:              taskQueue,
+		ScheduleToStartTimeout: timestamp.DurationFromSeconds(100),
+	}
+	_, _, err := s.matchingEngine.AddWorkflowTask(context.Background(), &addRequest)
+	if err != nil {
+		panic(err)
+	}
+}
+
+func (s *matchingEngineSuite) createPollWorkflowTaskRequestAndPoll(taskQueue *taskqueuepb.TaskQueue) {
+	result, err := s.matchingEngine.PollWorkflowTaskQueue(context.Background(), &matchingservice.PollWorkflowTaskQueueRequest{
+		NamespaceId: namespaceId,
+		PollRequest: &workflowservice.PollWorkflowTaskQueueRequest{
+			TaskQueue: taskQueue,
+			Identity:  "nobody",
+		},
+	}, metrics.NoopMetricsHandler)
+	if err != nil {
+		panic(err)
+	}
+	s.NotNil(result)
+	if len(result.TaskToken) == 0 {
+		panic("Empty poll returned")
+	}
+}
+
+// addWorkflowTasks is a helper that adds taskCount number of tasks for each numWorker
+func (s *matchingEngineSuite) addWorkflowTasks(concurrently bool, numWorkers int, taskCount int,
```

Comment:
I had that implementation in my previous push. However, what if I wanted to have 1 worker add and 1 worker poll concurrently? This was required in the last test case in the current commit, where 1 worker concurrently added 5 tasks and 1 poller concurrently polled 4 tasks, since I wanted to verify if the counter eventually read `1` or not. (TestLesserNumberOfPollersThanTasks)